lib/template.php

```diff
@@ -0,0 +1,213 @@
+<?php
+/**
+ * Dynamic template registration and overrides.
+ *
+ * @package gutenberg
+ */
+
+/**
+ * Replaces the output content of the template with the post content of a
+ * dynamic template, if it is the result of a dynamic template override.
+ *
+ * @see wp-includes/template-loader.php
+ *
+ * @param string $template The path or slug of the template to include.
+ */
+function gutenberg_template_include( $template ) {
+	// Assume if template override resulted in no dynamic template result, that
+	// original should be used.
+	if ( preg_match( '/\.php$/', $template ) ) {
+		return $template;
+	}
+
+	$template_post_query = new WP_Query( array(
+		'name'      => $template,
+		'post_type' => 'wp_template',
+	) );
+
+	// Abort and fall through to theme error messages.
+	if ( ! $template_post_query->have_posts() ) {
+		return;
+	}
+
+	$template_posts = $template_post_query->get_posts();
+	$template_post  = $template_posts[0];
+
+	echo apply_filters( 'the_content', $template_post->post_content );
+
+	// End of page lifecycle, abort to prevent theme error messages.
+	exit;
+}
+add_filter( 'template_include', 'gutenberg_template_include' );
+
+/**
+ * Given a template path, returns a template name with file extension removed.
+ *
+ * @param string $template_name The path of the template.
+ *
+ * @return string The name of the template with path extension removed.
+ */
+function gutenberg_map_template_slug( $template_name ) {
+	return preg_replace( '/\.php$/', '', $template_name );
+}
```

Comment:
It was an attempt to clarify that the function is meant to be used in the context of `array_map`, but I guess this isn't really a clarification that needs to be made. Maybe "get" instead of "map" ?